includes/razorpay-webhook.php

```diff
@@ -0,0 +1,179 @@
+     */
+    const ORDER_PAID  = 'order.paid';
+
+    function __construct()
+    {
+        $this->api = getRazorpayApiInstance();
+    }
+
+    /**
+     * Process a Razorpay Webhook. We exit in the following cases:
+     * - Successful processed
+     * - Exception while fetching the payment
+     *
+     * It passes on the webhook in the following cases:
+     * - invoice_id set in payment.authorized
+     * - Invalid JSON
+     * - Signature mismatch
+     * - Secret isn't setup
+     * - Event not recognized
+     */
+    public function process()
+    {
+        $post = file_get_contents('php://input');
+
+        $data = json_decode($post, true);
+
+        if (json_last_error() !== 0)
+        {
+            return;
+        }
+
+        $enabled = getSetting('enable_webhook');
+
+        if (isset($enabled) === true and
+            (empty($data['event']) === false))
+        {
+            if (isset($_SERVER['HTTP_X_RAZORPAY_SIGNATURE']) === true)
+            {
+                $razorpayWebhookSecret = getSetting('webhook_secret');
+
+                //
+                // If the webhook secret isn't set on wordpress, return
+                //
+                if (empty($razorpayWebhookSecret) === true)
+                {
+                    return;
+                }
+
+                try
+                {
+                    $this->api->utility->verifyWebhookSignature($post,
+                                                                $_SERVER['HTTP_X_RAZORPAY_SIGNATURE'],
+                                                                $razorpayWebhookSecret);
+                }
+                catch (Errors\SignatureVerificationError $e)
+                {
+                    $log = array(
+                        'message'   => $e->getMessage(),
+                        'data'      => $data,
+                        'event'     => 'razorpay.edd.signature.verify_failed'
+                    );
+
+                    error_log(json_encode($log));
+                    return;
+                }
+
+                switch ($data['event'])
+                {
+                    case self::ORDER_PAID:
+                        return $this->orderPaid($data);
+
+                    default:
+                        return;
+                }
+            }
+        }
+    }
+
+    /**
+     * Order Paid webhook
+     *
+     * @param array $data
+     */
+    protected function orderPaid(array $data)
+    {
+        // We don't process subscription/invoice payments here
+        if (isset($data['payload']['payment']['entity']['invoice_id']) === true)
+        {
+            return;
+        }
+
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['order']['entity']['notes']['edd_order_id'];
+
+        $order = new EDD_Payment(1491);
+
+        // If it is already marked as paid, ignore the event
+        if ($order->status === 'publish')
```

Comment:
Check should be only for initial state and process it if it's in initial state only. i.e when order is initiated